mantidproject/mantid
# Add Show Sample Shape to ADS right-click menu

**Description of work.**
 
Added right-click option to the ADS to allow the Sample Shape to be plotted.
This will plot all sample, container and component shapes attached to a workspace,
add a beam direction arrow in black and  if there is an attached crystal lattice (UB Matrix) then the real+reciprocal lattice vector arrows in rgb. These plots should also rename with the workspace, replot when the workspace is updated, in case SetSample or SetGoniometer was called, and close when the workspace is deleted.

Note two software behaviours that are not genuine problems but I came across in my own testing: Containers are not rotated by the goniometer. You may also notice that if a workspace has a container and you add just a sample then it will keep the container!

**Report to:** Pas and Spencer

**To test:**

- Review documentation MeshPlotHelp.rst which has been updated. It can be accessed via the `?` toolbar button on a SampleShapeplot (if you have built `docs-qthelp` and `docs-qtassistant`)
- Create a workspace with a sample and UB matrix:
```
cuboid = " \
<cuboid id='some-cuboid'> \
<height val='2.0'  /> \
<width val='2.0' />  \
<depth  val='0.2' />  \
<centre x='10.0' y='10.0' z='10.0'  />  \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

ws_shape = CreateSampleWorkspace()
SetGoniometer(ws, Axis0="120,0,0,1,1")
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid})
SetUB(ws, a=1, b=1, c=2, alpha=90, beta=90, gamma=60)
```
- Check you get the black beam arrows and the coloured lattice vector arrows.
- Run SetSample for one of the examples below (on the same workspace) and the plot should update in the same figurewindow as before.
- Right-click on the workspace and rename it, the plot should update with the new name.
- Delete the workspace and the plot should close.
- Create another workspace and Show the Sample Shape multiple times. Check they all the plots get renamed/close when the workspace is renamed/deleted.


- Use combinations of the below SetSample calls for general messing (more example code here: https://docs.mantidproject.org/nightly/algorithms/SetSample-v1.html#algm-setsample):


Just sample:
```
CreateWorkspace(OutputWorkspace="ws_shape", DataX=[1, 1], DataY=[2, 2])
merge_xml = ' \
<cylinder id="stick"> \
<centre-of-bottom-base x="-0.5" y="0.0" z="0.0" /> \
<axis x="1.0" y="0.0" z="0.0" />  \
<radius val="0.05" /> \
<height val="1.0" /> \
</cylinder> \
\
<sphere id="some-sphere"> \
<centre x="0.7"  y="0.0" z="0.0" /> \
<radius val="0.2" /> \
</sphere> \
\
<rotate-all x="90" y="-45" z="0" /> \
<algebra val="some-sphere (: stick)" /> \
'
SetSample("ws_shape", Geometry={'Shape': 'CSG', 'Value': merge_xml})

```
Sample and Container:
```
CreateWorkspace(OutputWorkspace=workspace_name, DataX=[1, 1], DataY=[2, 2])
SetSample(workspace_name,
          Geometry={'Shape': 'Cylinder', 'Height': 4.0,
                    'Radius': 2.0, 'Center': [0., 0., 0.]},
          Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6',
                    'NumberDensity': 0.1},
          ContainerGeometry={'Shape': 'HollowCylinder', 'Height': 4.0,
                             'InnerRadius': 2.0, 'OuterRadius': 2.3,
                             'Center': [0., 0., 0.]},
          ContainerMaterial={'ChemicalFormula': 'Al',
                             'NumberDensity': 0.01})
```
Just Container:
```
CreateWorkspace(OutputWorkspace=workspace_name, DataX=[1, 1], DataY=[2, 2])

SetSample(workspace_name,
          ContainerGeometry={'Shape': 'HollowCylinder', 'Height': 4.0,
                             'InnerRadius': 2.0, 'OuterRadius': 2.3,
                             'Center': [0., 0., 0.]},
          ContainerMaterial={'ChemicalFormula': 'Al',
                             'NumberDensity': 0.01})
```
Sample, Container and Components:
```
workspace = CreateWorkspace(OutputWorkspace="ws_shape", DataX=[1, 1], DataY=[2, 2])
LoadInstrument(Workspace=workspace, RewriteSpectraMap=True, InstrumentName="Pearl")
SetSample(workspace, Environment={'Name': 'Pearl'})
```

To test an instrument reference frame with pointing up Z and along beam X (e.g. POLREF) use this hack:
```
workspace = CreateWorkspace(OutputWorkspace="ws_shape", DataX=[1, 1], DataY=[2, 2])
LoadInstrument(Workspace=workspace, RewriteSpectraMap=True, InstrumentName="Polref")
SetSample(workspace, Environment={'Name': 'Pearl'})
```


Fixes #34563 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
